# Práctica 9. Clasificador Bayesiano

### Grupo 41
Alumnos:
- Óscar Rico Rodríguez
- Jia Hao Yang

### Tarea 1: Clasificación de mensajes como spam con Naive Bayes
Haciendo uso de los dos conjuntos de datos explicados en el apartado 9.2 se van a clasificar los mensajes como legítimos (ham) o spam. Para ello se hará uso de las implementaciones que hace la librería sklearn del clasificador Naive Bayes. Las clases están contenidas en el módulo naive_bayes y varían dependiendo de la hipótesis previas sobre la verosimilitud de cada característica (ec. 9.1.3). Las clases implementadas son  GaussianNB considera que la verosimilitud sigue una distribución normal N(µi ,σi), la clase MultinomialNB cuando las características se corresponden con el número de apariciones de un determinado valor como puede ser una palabra, la clase ComplementNB que es una variante del MultinomialNB cuando las clases están desbalanceadas o la clase BernoulliNB cuando las características son booleanas ya que supone una distribución de Bernoulli. De la misma forma que las clases de clasificadores vistas hasta ahora, estas clases que implementan diferentes versiones del clasificador Naive Bayes poseen los métodos fit y predict.

Así en esta tarea se deberá realizar lo siguiente:
- Dividir los conjuntos de entrenamiento en train (70%) y test (30%).
- Entrenar las diferentes implementaciones del clasificador Naive Bayes indicadas anteriormente.
- Obtener para los dos conjuntos de datos la tasa de acierto (accuracy).
- Obtener las matrices de confusión.
- Analizar y explicar los resultados obtenidos con cada clasificador en función de las medidas anteriores.

In [255]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB, ComplementNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

`spam_preprocesado_reales`

In [256]:
reales_accurracy = []
binarios_accurracy = []

In [257]:
spam_reales = pd.read_csv('spam_preprocesado_reales.csv')

In [258]:
spam_reales.head()

v1  150p  150ppm   16   18  1st  2nd  500  a100  a1000  ...  yes  \
0   ham   0.0     0.0  0.0  0.0  0.0  0.0  0.0   0.0    0.0  ...  0.0   
1   ham   0.0     0.0  0.0  0.0  0.0  0.0  0.0   0.0    0.0  ...  0.0   
2  spam   0.0     0.0  0.0  0.0  0.0  0.0  0.0   0.0    0.0  ...  0.0   
3   ham   0.0     0.0  0.0  0.0  0.0  0.0  0.0   0.0    0.0  ...  0.0   
4   ham   0.0     0.0  0.0  0.0  0.0  0.0  0.0   0.0    0.0  ...  0.0   

   yesterday  yet   yo  you  your  youre  yours  yourself  yup  
0        0.0  0.0  0.0  0.0   0.0    0.0    0.0       0.0  0.0  
1        0.0  0.0  0.0  0.0   0.0    0.0    0.0       0.0  0.0  
2        0.0  0.0  0.0  0.0   0.0    0.0    0.0       0.0  0.0  
3        0.0  0.0  0.0  0.0   0.0    0.0    0.0       0.0  0.0  
4        0.0  0.0  0.0  0.0   0.0    0.0    0.0       0.0  0.0  

[5 rows x 501 columns]

In [259]:
train, test = train_test_split(spam_reales, test_size=0.3)

In [260]:
gaussianNB = GaussianNB()
gaussianNB.fit(train.iloc[:, 1:], train['v1'])
gaussianNB.predict(test.drop('v1', axis=1))
accuracy = accuracy_score(test['v1'], gaussianNB.predict(test.drop('v1', axis=1)))
reales_accurracy.append(accuracy)
print(f'Accurracy GaussianNB: {accuracy:.3f}%')

Accurracy GaussianNB: 0.730%


In [261]:
gaussian_confusion = confusion_matrix(test['v1'], gaussianNB.predict(test.drop('v1', axis=1)))
print(gaussian_confusion)

[[996 440]
 [ 11 225]]


In [262]:
print(f'True positives: {gaussian_confusion[0][0]}')
print(f'False positives: {gaussian_confusion[0][1]}')
print(f'False negatives: {gaussian_confusion[1][0]}')
print(f'True negatives: {gaussian_confusion[1][1]}')

True positives: 996
False positives: 440
False negatives: 11
True negatives: 225


In [263]:
multinomialNB = MultinomialNB()
multinomialNB.fit(train.iloc[:, 1:], train['v1'])
multinomialNB.predict(test.drop('v1', axis=1))
accuracy = accuracy_score(test['v1'], multinomialNB.predict(test.drop('v1', axis=1)))
reales_accurracy.append(accuracy)
print(f'Accurracy MultinomialNB: {accuracy:.3f}%')

Accurracy MultinomialNB: 0.974%


In [264]:
multinomial_confusion = confusion_matrix(test['v1'], multinomialNB.predict(test.drop('v1', axis=1)))
print(multinomial_confusion)

[[1435    1]
 [  43  193]]


In [265]:
print(f'MultinomialNB True positives: {multinomial_confusion[0][0]}')
print(f'MultinomialNB False positives: {multinomial_confusion[0][1]}')
print(f'MultinomialNB False negatives: {multinomial_confusion[1][0]}')
print(f'MultinomialNB True negatives: {multinomial_confusion[1][1]}')

MultinomialNB True positives: 1435
MultinomialNB False positives: 1
MultinomialNB False negatives: 43
MultinomialNB True negatives: 193


In [266]:
bernoulliNB = BernoulliNB()
bernoulliNB.fit(train.iloc[:, 1:], train['v1'])
bernoulliNB.predict(test.drop('v1', axis=1))
accuracy = accuracy_score(test['v1'], bernoulliNB.predict(test.drop('v1', axis=1)))
reales_accurracy.append(accuracy)
print(f'Accurracy BernoulliNB: {accuracy:.3f}%')

Accurracy BernoulliNB: 0.978%


In [267]:
bernoulli_confusion = confusion_matrix(test['v1'], bernoulliNB.predict(test.drop('v1', axis=1)))
print(bernoulli_confusion)

[[1426   10]
 [  27  209]]


In [268]:
print(f'BernoulliNB True positives: {bernoulli_confusion[0][0]}')
print(f'BernoulliNB False positives: {bernoulli_confusion[0][1]}')
print(f'BernoulliNB False negatives: {bernoulli_confusion[1][0]}')
print(f'BernoulliNB True negatives: {bernoulli_confusion[1][1]}')

BernoulliNB True positives: 1426
BernoulliNB False positives: 10
BernoulliNB False negatives: 27
BernoulliNB True negatives: 209


In [269]:
complementNB = ComplementNB()
complementNB.fit(train.iloc[:, 1:], train['v1'])
complementNB.predict(test.drop('v1', axis=1))
accuracy = accuracy_score(test['v1'], complementNB.predict(test.drop('v1', axis=1)))
reales_accurracy.append(accuracy)
print(f'Accurracy ComplementNB: {accuracy:.3f}%')

Accurracy ComplementNB: 0.942%


In [270]:
complement_confusion = confusion_matrix(test['v1'], complementNB.predict(test.drop('v1', axis=1)))
print(complement_confusion)

[[1350   86]
 [  11  225]]


In [271]:
print(f'ComplemetNB True positives: {complement_confusion[0][0]}')
print(f'ComplemetNB False positives: {complement_confusion[0][1]}')
print(f'ComplemetNB False negatives: {complement_confusion[1][0]}')
print(f'ComplemetNB True negatives: {complement_confusion[1][1]}')

ComplemetNB True positives: 1350
ComplemetNB False positives: 86
ComplemetNB False negatives: 11
ComplemetNB True negatives: 225


`spam_preproceado_binario`

In [272]:
spam_binario = pd.read_csv('spam_preprocesado_binario.csv')

In [273]:
spam_binario.head()

v1  150p  150ppm  16  18  1st  2nd  500  a100  a1000  ...  yes  \
0   ham     0       0   0   0    0    0    0     0      0  ...    0   
1   ham     0       0   0   0    0    0    0     0      0  ...    0   
2  spam     0       0   0   0    0    0    0     0      0  ...    0   
3   ham     0       0   0   0    0    0    0     0      0  ...    0   
4   ham     0       0   0   0    0    0    0     0      0  ...    0   

   yesterday  yet  yo  you  your  youre  yours  yourself  yup  
0          0    0   0    0     0      0      0         0    0  
1          0    0   0    0     0      0      0         0    0  
2          0    0   0    0     0      0      0         0    0  
3          0    0   0    0     0      0      0         0    0  
4          0    0   0    0     0      0      0         0    0  

[5 rows x 501 columns]

In [274]:
train, test = train_test_split(spam_binario, test_size=0.3)

In [275]:
gaussianNB = GaussianNB()
gaussianNB.fit(train.iloc[:, 1:], train['v1'])
gaussianNB.predict(test.drop('v1', axis=1))
accuracy = accuracy_score(test['v1'], gaussianNB.predict(test.drop('v1', axis=1)))
binarios_accurracy.append(accuracy)
print(f'Accurracy GaussianNB: {accuracy:.3f}%')

Accurracy GaussianNB: 0.708%


In [276]:
gaussian_confusion = confusion_matrix(test['v1'], gaussianNB.predict(test.drop('v1', axis=1)))
print(gaussian_confusion)

[[975 478]
 [ 10 209]]


In [277]:
print(f'True positives: {gaussian_confusion[0][0]}')
print(f'False positives: {gaussian_confusion[0][1]}')
print(f'False negatives: {gaussian_confusion[1][0]}')
print(f'True negatives: {gaussian_confusion[1][1]}')

True positives: 975
False positives: 478
False negatives: 10
True negatives: 209


In [278]:
multinomialNB = MultinomialNB()
multinomialNB.fit(train.iloc[:, 1:], train['v1'])
multinomialNB.predict(test.drop('v1', axis=1))
accuracy = accuracy_score(test['v1'], multinomialNB.predict(test.drop('v1', axis=1)))
binarios_accurracy.append(accuracy)
print(f'Accurracy MultinomialNB: {accuracy:.3f}%')

Accurracy MultinomialNB: 0.975%


In [279]:
multinomial_confusion = confusion_matrix(test['v1'], multinomialNB.predict(test.drop('v1', axis=1)))
print(multinomial_confusion)

[[1435   18]
 [  24  195]]


In [280]:
print(f'MultinomialNB True positives: {multinomial_confusion[0][0]}')
print(f'MultinomialNB False positives: {multinomial_confusion[0][1]}')
print(f'MultinomialNB False negatives: {multinomial_confusion[1][0]}')
print(f'MultinomialNB True negatives: {multinomial_confusion[1][1]}')

MultinomialNB True positives: 1435
MultinomialNB False positives: 18
MultinomialNB False negatives: 24
MultinomialNB True negatives: 195


In [281]:
bernoulliNB = BernoulliNB()
bernoulliNB.fit(train.iloc[:, 1:], train['v1'])
bernoulliNB.predict(test.drop('v1', axis=1))
accuracy = accuracy_score(test['v1'], bernoulliNB.predict(test.drop('v1', axis=1)))
binarios_accurracy.append(accuracy)
print(f'Accurracy BernoulliNB: {accuracy:.3f}%')

Accurracy BernoulliNB: 0.978%


In [282]:
bernoulli_confusion = confusion_matrix(test['v1'], bernoulliNB.predict(test.drop('v1', axis=1)))
print(bernoulli_confusion)

[[1445    8]
 [  29  190]]


In [283]:
print(f'BernoulliNB True positives: {bernoulli_confusion[0][0]}')
print(f'BernoulliNB False positives: {bernoulli_confusion[0][1]}')
print(f'BernoulliNB False negatives: {bernoulli_confusion[1][0]}')
print(f'BernoulliNB True negatives: {bernoulli_confusion[1][1]}')

BernoulliNB True positives: 1445
BernoulliNB False positives: 8
BernoulliNB False negatives: 29
BernoulliNB True negatives: 190


In [284]:
complementNB = ComplementNB()
complementNB.fit(train.iloc[:, 1:], train['v1'])
complementNB.predict(test.drop('v1', axis=1))
accuracy = accuracy_score(test['v1'], complementNB.predict(test.drop('v1', axis=1)))
binarios_accurracy.append(accuracy)
print(f'Accurracy ComplementNB: {accuracy:.3f}%')

Accurracy ComplementNB: 0.950%


In [285]:
complement_confusion = confusion_matrix(test['v1'], complementNB.predict(test.drop('v1', axis=1)))
print(complement_confusion)

[[1386   67]
 [  17  202]]


In [286]:
print(f'ComplemetNB True positives: {complement_confusion[0][0]}')
print(f'ComplemetNB False positives: {complement_confusion[0][1]}')
print(f'ComplemetNB False negatives: {complement_confusion[1][0]}')
print(f'ComplemetNB True negatives: {complement_confusion[1][1]}')

ComplemetNB True positives: 1386
ComplemetNB False positives: 67
ComplemetNB False negatives: 17
ComplemetNB True negatives: 202


In [287]:
models = ['GaussianNB', 'MultinomialNB', 'BernoulliNB', 'ComplementNB']
data = {'Model': models, 'Accuracy Reales': reales_accurracy, 'Accuracy Binarios': binarios_accurracy}
acc_df = pd.DataFrame(data)
acc_df

Model  Accuracy Reales  Accuracy Binarios
0     GaussianNB         0.730263           0.708134
1  MultinomialNB         0.973684           0.974880
2    BernoulliNB         0.977871           0.977871
3   ComplementNB         0.941986           0.949761

`BernoulliNB`

Con `BernoulliNB` este clasificador es estupendo para casos Binarios. El clasificador asume que las características del conjunto de datos sigue una distribución de Bernoulli. En nuestro caso y en el teoría, este clasificador sería muy bueno para el conjunto de datos `spam_preprocesado_binario` y es cierto, tiene una precisión muy muy alta. Pero sorprendentemente, este también da muy buenos resultados en el conjunto de datos `spam_preprocesado_binario`, que este conjunto no es binario y solo recoge las frecuncias de las palabras.

`ComplementNB`

El clasificador `ComplementNB` es bastante parecido al clasificador `MultinomialNB` pero a diferencia de este, este sirve para conjuntos de datos desequilibrados. Este quiere decir, que este es más efectivo en conjuntos de datos, donde una característica es más frecuente que otra. Esta sirve también para la clasificación de palabras y texto ya que en el lenguaje natural, la distribución de palabras que utilizamos no es equitativa para todas las palabras, de ahí su alta precisión.

`MultinomialNB`

Con este clasificador, podemos ver un salto de mejora en cuanto a la precisión de la predicción con respecto al clasificador anterior. `GaussianNB` trata de asumir una distribución normal para la características, en cambio, `MultinomialNB`, asume una distribución multinomial discreta. Esta es muy buena para el conteo de muestras discretas, como es nuestro caso. En nuestro caso realizamos tanto un conteo por frecuencias de las palabras y una detección binaria si dicha palabra está en el contenido del texto.

`GaussianNB`

En la teoría, puesto sobre el papel, `GaussianNB` es el mejor para este tipo de clasificaciones, sobre todo el caso de los reales. Pero como podemos observar, este clasificador es bastante malo para este caso y peor aún, para el conjunto de datos binarios.

### Tarea 2: Comparación de clasificación de mensajes spam
Para poder comprobar el rendimiento del clasificador Naive Bayes en la clasificación de mensajes SMS como legítimos o spam, se debería comparar los resultados con otros clasificadores. Por tanto, en esta tarea se realizará la clasificación de los dos conjuntos anteriores con diferentes clasificadores de los utilizados hasta ahora en las prácticas y de esta manera comprobar la bondad del clasificador Naive Bayes.

Así en esta tarea se deberá realizar lo siguiente:
- Dividir los conjuntos de entrenamiento en train (70%) y test (30%).
- Entrenar los clasificadores vistos en las prácticas anteriores (Perceptron, Regresión Logística, SVM, Vecinos más Cercanos).
- Mostrar en una tabla para cada conjunto de datos, los valores de la tasa de acierto (accuracy).
- Analizar y explicar los resultados obtenidos.

In [288]:
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import Perceptron
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

Cargamos los datos de ambos csv

In [289]:
spam_reales_2 = pd.read_csv('spam_preprocesado_reales.csv', sep=',')
spam_binario_2 = pd.read_csv('spam_preprocesado_binario.csv', sep=',')

Cambiamos el nombre de la variable 'v1' a 'result', ya que no es descriptiva

In [290]:
spam_reales_2.rename(columns={'v1':'result'}, inplace=True)
spam_binario_2.rename(columns={'v1':'result'}, inplace=True)

##### Dividir los conjuntos de entrenamiento en train (70%) y test (30%).

In [291]:
x_train_real, x_test_real, y_train_real, y_test_real = train_test_split(spam_reales_2.drop('result', axis=1), spam_reales_2['result'], test_size=0.3, random_state=0)
x_train_bin, x_test_bin, y_train_bin, y_test_bin = train_test_split(spam_binario_2.drop('result', axis=1), spam_binario_2['result'], test_size=0.3, random_state=0)

##### Entrenar los clasificadores vistos en las prácticas anteriores (Perceptron, Regresión Logística, SVM, Vecinos más Cercanos).

In [292]:
data = [[x_train_real, x_test_real, y_train_real, y_test_real, 'Real'], [x_train_bin, x_test_bin, y_train_bin, y_test_bin, 'Binario']]

In [293]:
results_perceptron = pd.DataFrame()
for list in data:
    perceptron = Perceptron()
    perceptron.fit(list[0], list[2])
    yhat = perceptron.predict(list[1])
    results_perceptron[f'Acurracy {list[4]}'] = [accuracy_score(list[3], yhat)]
results_perceptron.index = ['Perceptron']

In [294]:
results_logistic = pd.DataFrame()
for list in data:
    logistic = LogisticRegression()
    logistic.fit(list[0], list[2])
    yhat = logistic.predict(list[1])
    results_logistic[f'Acurracy {list[4]}'] = [accuracy_score(list[3], yhat)]
results_logistic.index = ['Logistic']

In [295]:
results_svc = pd.DataFrame()
c_list = [0.01, 0.1, 1, 10]
type_list = ['linear', 'poly', 'rbf']
for list in data:
    accuracy_list = []
    index = []
    for i in type_list:
        for j in c_list:
            svc = SVC(kernel=i, C=j)
            svc.fit(list[0], list[2])
            yhat = svc.predict(list[1])
            accuracy_list.append(accuracy_score(list[3], yhat))
            index.append(f'SVC {i} {j}')
    results_svc[f'Acurracy {list[4]}'] = accuracy_list
    results_svc.index = index

In [296]:
results_knn = pd.DataFrame()
for list in data:
    accuracy_list = []
    index = []
    for i in range(1,11):
        knn = KNeighborsClassifier(n_neighbors=i)
        knn.fit(list[0], list[2])
        yhat = knn.predict(list[1])
        accuracy_list.append(accuracy_score(list[3], yhat))
        index.append(f'KNN {i}')
    results_knn[f'Acurracy {list[4]}'] = accuracy_list
    results_knn.index = index

##### Mostrar en una tabla para cada conjunto de datos, los valores de la tasa de acierto (accuracy).

In [297]:
result = pd.concat([results_perceptron,results_logistic,results_svc,results_knn])

In [298]:
result

Acurracy Real  Acurracy Binario
Perceptron            0.975478          0.972488
Logistic              0.963517          0.979665
SVC linear 0.01       0.857656          0.946770
SVC linear 0.1        0.946770          0.975478
SVC linear 1          0.971292          0.979067
SVC linear 10         0.977273          0.962919
SVC poly 0.01         0.857656          0.859450
SVC poly 0.1          0.858254          0.914474
SVC poly 1            0.958134          0.955742
SVC poly 10           0.958134          0.962919
SVC rbf 0.01          0.857656          0.857656
SVC rbf 0.1           0.892344          0.876196
SVC rbf 1             0.975478          0.974880
SVC rbf 10            0.979665          0.977273
KNN 1                 0.955144          0.949761
KNN 2                 0.940789          0.934211
KNN 3                 0.939593          0.938397
KNN 4                 0.924641          0.919856
KNN 5                 0.924641          0.925837
KNN 6                 0.909689          0.910287
KNN 7                 0.909689          0.912081
KNN 8                 0.898325          0.904904
KNN 9                 0.898325          0.904904
KNN 10                0.890550          0.898325

Mostramos los mejores valores para el csv de spam real

In [299]:
result.sort_values(by='Acurracy Real', ascending=False)

Acurracy Real  Acurracy Binario
SVC rbf 10            0.979665          0.977273
SVC linear 10         0.977273          0.962919
Perceptron            0.975478          0.972488
SVC rbf 1             0.975478          0.974880
SVC linear 1          0.971292          0.979067
Logistic              0.963517          0.979665
SVC poly 1            0.958134          0.955742
SVC poly 10           0.958134          0.962919
KNN 1                 0.955144          0.949761
SVC linear 0.1        0.946770          0.975478
KNN 2                 0.940789          0.934211
KNN 3                 0.939593          0.938397
KNN 4                 0.924641          0.919856
KNN 5                 0.924641          0.925837
KNN 6                 0.909689          0.910287
KNN 7                 0.909689          0.912081
KNN 8                 0.898325          0.904904
KNN 9                 0.898325          0.904904
SVC rbf 0.1           0.892344          0.876196
KNN 10                0.890550          0.898325
SVC poly 0.1          0.858254          0.914474
SVC rbf 0.01          0.857656          0.857656
SVC poly 0.01         0.857656          0.859450
SVC linear 0.01       0.857656          0.946770

Mostramos los mejores valores para el conjunto de datos Binario

In [300]:
result.sort_values(by='Acurracy Binario', ascending=False)

Acurracy Real  Acurracy Binario
Logistic              0.963517          0.979665
SVC linear 1          0.971292          0.979067
SVC rbf 10            0.979665          0.977273
SVC linear 0.1        0.946770          0.975478
SVC rbf 1             0.975478          0.974880
Perceptron            0.975478          0.972488
SVC poly 10           0.958134          0.962919
SVC linear 10         0.977273          0.962919
SVC poly 1            0.958134          0.955742
KNN 1                 0.955144          0.949761
SVC linear 0.01       0.857656          0.946770
KNN 3                 0.939593          0.938397
KNN 2                 0.940789          0.934211
KNN 5                 0.924641          0.925837
KNN 4                 0.924641          0.919856
SVC poly 0.1          0.858254          0.914474
KNN 7                 0.909689          0.912081
KNN 6                 0.909689          0.910287
KNN 9                 0.898325          0.904904
KNN 8                 0.898325          0.904904
KNN 10                0.890550          0.898325
SVC rbf 0.1           0.892344          0.876196
SVC poly 0.01         0.857656          0.859450
SVC rbf 0.01          0.857656          0.857656

##### Analizar y explicar los resultados obtenidos.

`Perceptron`

En esta prueba, el Perceptrón ha demostrado ser un clasificador sólido, obteniendo el tercer mejor rendimiento en el conjunto de datos real y el sexto en el conjunto de datos binario, con una precisión superior al 97% en ambos casos. Esto es sorprendente considerando que se trata de un clasificador simple con un tiempo de ejecución muy reducido.

`Logistic Regression`

La Regresión Logística se ha destacado como la mejor opción para clasificar el conjunto de datos binarios, con una precisión cercana al 98%, y como la sexta mejor opción para clasificar el conjunto de datos reales. Además, su tiempo de ejecución es bastante corto y al tratarse de un clasificador simple, ha obtenido resultados espectaculares. Por lo tanto, es altamente recomendable si se necesita resultados rápidos y no se dispone de mucho poder de computación.

`SVC`

El clasificador SVC ha demostrado ser una buena opción, siempre y cuando se sepa cómo ajustar sus parámetros. Al usar un kernel lineal con C=1, ha obtenido el segundo mejor rendimiento para el conjunto de datos binarios y, al usar un kernel rbf y C=10, ha sido el mejor clasificador para los datos reales, ambos con una precisión cercana al 98%. Por lo tanto, es altamente recomendable si se busca un único clasificador que proporcione buenos resultados en ambos conjuntos de datos, se dispone de suficiente poder de computación y se ajustan bien sus parámetros.

`KNN`

El clasificador KNN ha obtenido resultados muy deficientes en esta prueba en comparación con los demás clasificadores. En casi ninguno de los conjuntos de datos logra ubicarse en el Top 10, y en el caso en que lo hace, es gracias al clasificador NN, donde solo se comprueba el vecino más cercano (k=1). Por lo tanto, este clasificador no es recomendable para este tipo de ejercicios.

##### Comparación con Ejercicio 1

Por lo que podemos observar, todos los clasificadores del módulo `naive_bayes` es bastante bueno con la excepción del `GaussianNB`. Con los clasificadores que hemos utilizado aquí, todos salvo el K-NN dan resultados muy similires y bastantes buenos como los de `naive_bayes`.